src/app/address/controllers/address/nonUKAddress.js

```diff
@@ -0,0 +1,124 @@
+const BaseController = require("hmpo-form-wizard").Controller;
+const countryList = require("../../data/countries.json");
+const {
+  validateBuildingAddress,
+} = require("../../validators/nonUKAddressValidator");
+
+class NonUKAddressController extends BaseController {
+  getValues(req, res, callback) {
+    super.getValues(req, res, (err, values) => {
+      values.addressCountryName = this.mapToName(
+        req.translate,
+        req.sessionModel.get("country")
+      );
+
+      Object.entries(req?.form?.errors || {})?.map(
+        ([fieldName, validationAttribute]) =>
+          (values[`${fieldName}InValid`] = this.getErrorMessage(
+            req.translate,
+            validationAttribute.key,
+            validationAttribute.type
+          ))
+      );
+
+      values.buildAddressErrorMessage = this.hasBuildingEmptyErrors(req) && {
+        text: req.translate("validation.apartmentOrBuildingNumberOrName"),
+        visuallyHiddenText: "error",
+      };
+
+      callback(err, values);
+    });
+  }
+
+  validateFields(req, res, callback) {
+    const formFields = req.form.options.fields;
+    const apartment = "nonUKAddressApartmentNumber";
+    const number = "nonUKAddressBuildingNumber";
+    const name = "nonUKAddressBuildingName";
+
+    const fields = {
+      [apartment]: req.body.nonUKAddressApartmentNumber,
+      [number]: req.body.nonUKAddressBuildingNumber,
+      [name]: req.body.nonUKAddressBuildingName,
+    };
+
+    const buildingAddress = Object.entries(fields);
+
+    buildingAddress.every(fieldIsEmpty) &&
+      this.defaultToFirstField(formFields, apartment, req);
+
+    super.validateFields(req, res, callback);
+  }
+
+  async saveValues(req, res, callback) {
+    super.saveValues(req, res, () => {
+      const addressCountry = req.sessionModel.get("country");
+      const address = this.buildAddress(req.body, addressCountry);
+      req.sessionModel.set("address", address);
+
+      callback();
+    });
+  }
+
+  buildAddress(
+    {
+      nonUKAddressApartmentNumber,
+      nonUKAddressBuildingNumber,
+      nonUKAddressBuildingName,
+      nonUKAddressStreetName,
+      nonUKAddressLocality,
+      nonUKAddressPostalCode,
+      nonUKAddressRegion,
+      nonUKAddressYearFrom,
+    } = undefined,
+    addressCountry
+  ) {
+    const yearFrom = nonUKAddressYearFrom
+      ? new Date(nonUKAddressYearFrom).toISOString().split("T")[0]
+      : null;
+
+    return {
+      addressRegion: nonUKAddressRegion,
+      addressLocality: nonUKAddressLocality,
+      streetName: nonUKAddressStreetName,
+      postalCode: nonUKAddressPostalCode,
+      buildingNumber: nonUKAddressBuildingNumber,
+      buildingName: nonUKAddressBuildingName,
+      subBuildingName: nonUKAddressApartmentNumber,
+      validFrom: yearFrom,
+      addressCountry,
+    };
+  }
+
+  mapToName(translate, countryCode) {
+    return translate(
+      countryList.find((country) => country.value === countryCode).key
+    );
```

Comment:
I think this is going to be needed on the check details screen too (@mikebeeby correct me if I'm wrong) so maybe we also want to put this somewhere shared?